zitadel/user_grant/funcs.go

```diff
@@ -125,3 +126,169 @@ func read(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagn
 	d.SetId(grant.GetId())
 	return nil
 }
+
+func readDS(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	tflog.Info(ctx, "started read")
+
+	clientinfo, ok := m.(*helper.ClientInfo)
+
+	if !ok {
+
+		return diag.Errorf("failed to get client")
+
+	}
+
+	client, err := helper.GetManagementClient(ctx, clientinfo)
+
+	if err != nil {
+
+		return diag.FromErr(err)
+
+	}
+
+	resp, err := client.GetUserGrantByID(helper.CtxWithOrgID(ctx, d), &management.GetUserGrantByIDRequest{
+
+		GrantId: helper.GetID(d, grantIDVar),
+
+		UserId: d.Get(UserIDVar).(string),
+	})
+
+	if err != nil && helper.IgnoreIfNotFoundError(err) == nil {
+
+		d.SetId("")
+
+		return nil
+
+	}
+
+	if err != nil {
+
+		return diag.Errorf("failed to get user grant")
+
+	}
+
+	grant := resp.GetUserGrant()
+
+	set := map[string]interface{}{
+
+		UserIDVar: grant.GetUserId(),
+
+		RoleKeysVar: grant.GetRoleKeys(),
+
+		userNameVar: grant.GetUserName(),
+
+		roleStatusVar: grant.GetState().String(),
+
+		projectNameVar: grant.GetProjectName(),
+
+		emailVar: grant.Email,
+
+		nameVar: grant.DisplayName,
+	}
+
+	if grant.GetProjectId() != "" {
+
+		set[projectIDVar] = grant.GetProjectId()
+
+	}
+
+	if grant.GetProjectGrantId() != "" {
+
+		set[projectGrantIDVar] = grant.GetProjectGrantId()
+
+	}
+
+	for k, v := range set {
+
+		if err := d.Set(k, v); err != nil {
+
+			return diag.Errorf("failed to set %s of usergrant: %v", k, err)
+
+		}
+
+	}
+
+	d.SetId(grant.GetId())
+
+	return nil
+
+}
+
+func list(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	tflog.Info(ctx, "started list")
+
+	projectName := d.Get(projectNameVar).(string)
+
+	clientinfo, ok := m.(*helper.ClientInfo)
+
+	if !ok {
+
+		return diag.Errorf("failed to get client")
+
+	}
+
+	client, err := helper.GetManagementClient(ctx, clientinfo)
+
+	if err != nil {
+
+		return diag.FromErr(err)
+
+	}
+
+	req := &management.ListUserGrantRequest{}
+
+	req.Queries = append(req.Queries, &user.UserGrantQuery{
+
+		Query: &user.UserGrantQuery_ProjectNameQuery{
+
+			ProjectNameQuery: &user.UserGrantProjectNameQuery{
+
+				ProjectName: projectName,
```

Comment:
Why is only the projectName used as a query? It is even listed as computed, I would rather use other identifiers first, and maybe provide the name later.

Reply:
i am using projectName as the query because that is my current use case, i need to lookup the user grants by project name.  Feel free to make it more generic or point me to an example of how you want it done.

Reply:
Sounds good to me for the start, normally I would at least add the identifier of the project itself as a base query.
Have a look at the application_saml resouce as an example.